03-task-manager/starter/controllers/tasks.js

```diff
@@ -1,25 +1,72 @@
 const Task = require('../models/task');
-const getAllTasks = (req, res) => {
-    res.send('get all tasks');
-}
+const asyncWrapper = require('../middleware/async');
+const { createCustomError } = require('../errors/custom-error');
 
-const createTask = async(req, res) => {
+const getAllTasks = asyncWrapper( async (req, res) => {    
+    const tasks = await Task.find({});
+    res.status(200).json({tasks});
+    // res.status(200).json({tasks, amount: tasks.length});
+    // res.status(200).json({status: "success", data: {tasks,nbHits: tasks.length} });     
+})
+
+const createTask = asyncWrapper(async(req, res) => {   
     const task = await Task.create(req.body);
-    res.status(201).json({task});
-}
-const getTask = (req, res) => {
-    res.json({id:req.params.id});
-}
-const updateTask = (req, res) => {
-    res.send('update task');
-}
-const deleteTask = (req, res) => {
-    res.send('delete task');
-}
+    res.status(201).json({task});    
+})
+
+const getTask = asyncWrapper(async (req, res, next) => {   
+    const {id: taskID} = req.params;
+    const task = await Task.findOne({_id: taskID});
```

Comment:
Nitpick (deson't require changes):
I'd recommend checking if `req.params.id` or **any other form of stringified id** is a valid ObjectID by using mongoose's [isValidObjectId utility function](https://mongoosejs.com/docs/5.x/docs/api.html#mongoose_Mongoose-isValidObjectId) BEFORE using it within a `Task` method. Otherwise, if you supply an invalid id, then... for example, `Task.findOne`, will throw a casting error:
```
{
  err: CastError: Cast to ObjectId failed for value "1" (type string) at path "_id" for model "Task"
      at model.Query.exec (/Users/mattcarlotta/Documents/node-express-course/03-task-manager/starter/node_modules/.pnpm/mongoose@5.13.22/node_modules/mongoose/lib/query.js:4498:21)
      at Query.then (/Users/mattcarlotta/Documents/node-express-course/03-task-manager/starter/node_modules/.pnpm/mongoose@5.13.22/node_modules/mongoose/lib/query.js:4592:15) {
    messageFormat: undefined,
    stringValue: '"1"',
    kind: 'ObjectId',
    value: '1',
    path: '_id',
    reason: Error: Argument passed in must be a single String of 12 bytes or a string of 24 hex characters
        at new ObjectID (/Users/mattcarlotta/Documents/node-express-course/03-task-manager/starter/node_modules/.pnpm/bson@1.1.6/node_modules/bson/lib/bson/objectid.js:59:11)
        at castObjectId (/Users/mattcarlotta/Documents/node-express-course/03-task-manager/starter/node_modules/.pnpm/mongoose@5.13.22/node_modules/mongoose/lib/cast/objectid.js:25:12)
        at ObjectId.cast (/Users/mattcarlotta/Documents/node-express-course/03-task-manager/starter/node_modules/.pnpm/mongoose@5.13.22/node_modules/mongoose/lib/schema/objectid.js:246:12)
        at SchemaType.applySetters (/Users/mattcarlotta/Documents/node-express-course/03-task-manager/starter/node_modules/.pnpm/mongoose@5.13.22/node_modules/mongoose/lib/schematype.js:1123:12)
        at SchemaType._castForQuery (/Users/mattcarlotta/Documents/node-express-course/03-task-manager/starter/node_modules/.pnpm/mongoose@5.13.22/node_modules/mongoose/lib/schematype.js:1601:15)
        at SchemaType.castForQuery (/Users/mattcarlotta/Documents/node-express-course/03-task-manager/starter/node_modules/.pnpm/mongoose@5.13.22/node_modules/mongoose/lib/schematype.js:1591:15)
        at SchemaType.castForQueryWrapper (/Users/mattcarlotta/Documents/node-express-course/03-task-manager/starter/node_modules/.pnpm/mongoose@5.13.22/node_modules/mongoose/lib/schematype.js:1568:20)
        at cast (/Users/mattcarlotta/Documents/node-express-course/03-task-manager/starter/node_modules/.pnpm/mongoose@5.13.22/node_modules/mongoose/lib/cast.js:332:32)
        at Query.cast (/Users/mattcarlotta/Documents/node-express-course/03-task-manager/starter/node_modules/.pnpm/mongoose@5.13.22/node_modules/mongoose/lib/query.js:4937:12)
        at Query._castConditions (/Users/mattcarlotta/Documents/node-express-course/03-task-manager/starter/node_modules/.pnpm/mongoose@5.13.22/node_modules/mongoose/lib/query.js:1975:10)
        at model.Query.<anonymous> (/Users/mattcarlotta/Documents/node-express-course/03-task-manager/starter/node_modules/.pnpm/mongoose@5.13.22/node_modules/mongoose/lib/query.js:2232:8)
        at model.Query._wrappedThunk [as _findOne] (/Users/mattcarlotta/Documents/node-express-course/03-task-manager/starter/node_modules/.pnpm/mongoose@5.13.22/node_modules/mongoose/lib/helpers/query/wrapThunk.js:16:8)
        at /Users/mattcarlotta/Documents/node-express-course/03-task-manager/starter/node_modules/.pnpm/kareem@2.3.2/node_modules/kareem/index.js:370:33
        at process.processTicksAndRejections (node:internal/process/task_queues:77:11),
    valueType: 'string'
  }
}
```

For example (this would be true for ALL your controllers that use an id):
```js
const mongoose = require("mongoose");
const Task = require("../models/Task");
const asyncWrapper = require("../middleware/async");
const { createCustomError } = require("../errors/custom-error");

const getTask = asyncWrapper(async (req, res, done) => {
    const { id: taskID } = req.params;

    if (!mongoose.isValidObjectId(taskID)) {
        return done(createCustomError("Invalid task id", 400));
    }

    const task = await Task.findOne({ _id: taskID });

    if (!task) {
        return done(createCustomError(`No task with Id : ${taskID}`, 404));
    }

    res.status(200).json({ task });
});
```

Reply:
Thank you for showing me another, more clear version! 
Reading mongoose documentation and searching about mongoose.isValidObjectId method, I found out this phrase "mongoose.isValidObjectId(new mongoose.Types.ObjectId());"   What is better to use? mongoose.isValidObjectId   or  mongoose.Types.ObjectId())?

Reply:
I believe they’re both the same.